pkg/agent/openflow/pipeline.go

```diff
@@ -387,19 +387,37 @@ var (
 )
 
 type OFEntryOperations interface {
-	Add(flow binding.Flow) error
-	Modify(flow binding.Flow) error
-	Delete(flow binding.Flow) error
-	AddAll(flows []binding.Flow) error
-	ModifyAll(flows []binding.Flow) error
-	BundleOps(adds []binding.Flow, mods []binding.Flow, dels []binding.Flow) error
-	DeleteAll(flows []binding.Flow) error
+	AddAll(flowMessages []ofctrl.OpenFlowModMessage) error
+	ModifyAll(flowMessages []ofctrl.OpenFlowModMessage) error
+	BundleOps(adds, mods, dels []ofctrl.OpenFlowModMessage) error
+	DeleteAll(flowMessages []ofctrl.OpenFlowModMessage) error
 	AddOFEntries(ofEntries []binding.OFEntry) error
 	ModifyOFEntries(ofEntries []binding.OFEntry) error
 	DeleteOFEntries(ofEntries []binding.OFEntry) error
 }
 
-type flowCache map[string]binding.Flow
+type cachedFlow struct {
+	key        string
+	message    ofctrl.OpenFlowModMessage
+	table      binding.Table
+	isDropFlow bool
```

Comment:
I reviewed the current usage of key, table, and isDropFlow. The first one seems only used by methods added for antctl, which can be easily changed to return flows directly instead. The latter two are used only by `flowsToTrace`, which could actually get the same information from structs like conjunction or context. So perhaps we don't need to create this struct?

Reply:
Removed the fields according to the offline discussion, and publish the functions to dynamically generate Flow string or match string from a FlowModification mesage. Could you review it again? @tnqn